packages/react-shadow/src/renderers/createShadowDOMRenderer.ts

```diff
@@ -0,0 +1,204 @@
+    );
+
+    if (mediaElements.length) {
+      styleSheets = mediaElements;
+      comparer = (sheet) =>
+        renderer.compareMediaQueries(
+          styleSheet.metadata['m'] as string,
+          sheet.metadata['m'] as string
+        );
+    }
+  }
+
+  const length = styleSheets.length;
+  let index = length - 1;
+
+  while (index >= 0) {
+    const styleElement = styleSheets[index];
+
+    if (comparer(styleElement) > 0) {
+      return styleElement;
+    }
+
+    index--;
+  }
+
+  if (length > 0) {
+    return styleSheets[0];
+  }
+
+  return null;
+}
+
+function getCSSStyleSheetForBucket(
+  cssSheetsCache: Record<string, ExtendedCSSStyleSheet>,
+
+  bucketName: StyleBucketName,
+  metadata: Record<string, unknown> = {},
+
+  onStyleSheetInsert: (stylesheet: ExtendedCSSStyleSheet) => void
+): ExtendedCSSStyleSheet {
+  const isMediaBucket = bucketName === 'm';
+  const styleSheetKey: StyleBucketName | string = isMediaBucket
+    ? ((bucketName + metadata['m']) as string)
+    : bucketName;
+
+  if (!cssSheetsCache[styleSheetKey]) {
+    const styleSheet = new CSSStyleSheet() as ExtendedCSSStyleSheet;
+
+    styleSheet.bucketName = bucketName;
+    styleSheet.metadata = metadata;
+
+    cssSheetsCache[styleSheetKey] = styleSheet;
+    onStyleSheetInsert(styleSheet);
+  }
+
+  return cssSheetsCache[styleSheetKey];
+}
+
+function insertBefore<T extends CSSStyleSheet | ExtendedCSSStyleSheet>(
+  arr: T[],
+  sheetToInsert: T,
+  targetSheet: T | null
+): T[] {
+  if (targetSheet === null) {
+    return [...arr, sheetToInsert];
+  }
+
+  const index = arr.indexOf(targetSheet);
+
+  return [...arr.slice(0, index + 1), sheetToInsert, ...arr.slice(index + 1)];
+}
+
+export function createShadowDOMRenderer(shadowRoot: ShadowRoot) {
+  if (!SUPPORTS_CONSTRUCTABLE_STYLESHEETS) {
+    return createFallbackRenderer(shadowRoot) as GriffelRenderer & {
+      adoptedStyleSheets?: never;
+    };
+  }
+
+  const cssSheetsCache: Record<string, ExtendedCSSStyleSheet> = {};
+  const renderer: GriffelShadowDOMRenderer = {
+    id: `@fluentui-contrib/react-shadow:${rendererId++}`,
+
+    adoptedStyleSheets: [],
+    insertionCache: {},
+    stylesheets: {},
+
+    compareMediaQueries: (a: string, b: string) => (a < b ? -1 : a > b ? 1 : 0),
+    insertCSSRules(cssRules) {
+      for (const [_styleBucketName, cssBucketEntries] of Object.entries(
+        cssRules
+      )) {
+        const styleBucketName = _styleBucketName as StyleBucketName;
+
+        for (let i = 0, l = cssBucketEntries.length; i < l; i++) {
+          const [ruleCSS, metadata] = normalizeCSSBucketEntry(
+            cssBucketEntries[i]
+          );
+          const sheet = getCSSStyleSheetForBucket(
+            cssSheetsCache,
+
+            styleBucketName,
+            metadata,
+
+            (styleSheet) => {
+              const targetStyleSheet = findInsertionPoint(renderer, styleSheet);
+
+              renderer.adoptedStyleSheets = insertBefore(
```

Comment:
if griffel stylesheets could somehow keep references to each other, there's no need to maintain a copy of the relative order of adopted style sheets